File: frontend/src/types/asset.ts
// 资产相关类型定义

import { Organization, User } from './user';
import { Status, BaseQueryParams } from './common';

// 租户
export interface Tenant {
  id: number;
  tenantCode: string;
  tenantName: string;
  tenantType: 'individual' | 'company' | 'government' | 'organization';
  contactPerson: string;
  contactPhone: string;
  contactEmail?: string;
  companyName?: string;
  businessLicense?: string;
  taxNumber?: string;
  address?: string;
  industry?: string;
  businessScope?: string;
  creditRating?: 'AAA' | 'AA' | 'A' | 'BBB' | 'BB' | 'B' | 'C';
  status: 'active' | 'inactive' | 'blacklisted';
  createdAt: string;
  updatedAt: string;
  leases?: Lease[];
}

// 租赁合同
export interface Lease {
  id: number;
  leaseCode: string;
  roomId: number;
  room?: Room;
  tenantId: number;
  tenant?: Tenant;
  leaseType: 'rent' | 'sublease' | 'license';
  startDate: string;
  endDate: string;
  monthlyRent: number;
  deposit: number;
  managementFee?: number;
  utilities?: string[];
  paymentCycle: 'monthly' | 'quarterly' | 'semi-annual' | 'annual';
  paymentMethod: 'cash' | 'transfer' | 'check' | 'online';
  autoRenewal: boolean;
  renewalPeriod?: number;
  status: 'draft' | 'active' | 'expired' | 'terminated' | 'suspended';
  signedDate?: string;
  contractFile?: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
  payments?: Payment[];
}

// 付款记录
export interface Payment {
  id: number;
  leaseId: number;
  lease?: Lease;
  paymentType: 'rent' | 'deposit' | 'management' | 'utility' | 'penalty' | 'refund';
  amount: number;
  dueDate: string;
  paidDate?: string;
  paidAmount?: number;
  paymentMethod?: 'cash' | 'transfer' | 'check' | 'online';
  status: 'pending' | 'paid' | 'overdue' | 'partial' | 'cancelled';
  receipt?: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

// 资产信息
export interface Asset {
  id: number;
  assetCode: string;
  assetName: string;
  streetId?: number;
  streetName?: string;
  address?: string;
  longitude?: number;
  latitude?: number;
  landNature?: string;
  totalArea?: number;
  rentableArea?: number;
  assetTags?: string[];
  assetImages?: string[];
  description?: string;
  useDate?: string;
  status: Status;
  buildings?: Building[];
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

// 楼宇信息
export interface Building {
  id: number;
  assetId: number;
  assetName?: string;
  buildingCode: string;
  buildingName: string;
  buildingShortName?: string;
  buildingType: string;
  investor?: string;
  location?: string;
  totalArea?: number;
  rentableArea?: number;
  buildingHeight?: number;
  floorCount?: number;
  undergroundFloors?: number;
  constructionYear?: string;
  elevatorCount?: number;
  parkingSpaces?: number;
  greenRate?: number;
  propertyCompany?: string;
  propertyPhone?: string;
  buildingTags?: string[];
  buildingImages?: string[];
  features?: string[];
  description?: string;
  buildDate?: string;
  status: Status;
  floors?: Floor[];
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

// 楼层信息
export interface Floor {
  id: number;
  buildingId: number;
  buildingName?: string;
  floorCode: string;
  floorName: string;
  floorNumber: number;
  totalArea?: number;
  rentableArea?: number;
  rentedArea?: number;
  floorType?: string;
  floorPlan?: string;
  roomCount?: number;
  rentedRoomCount?: number;
  avgRentPrice?: number;
  occupancyRate?: number;
  remark?: string;
  status: Status;
  rooms?: Room[];
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

// 房间信息
export interface Room {
  id: number;
  floorId: number;
  floorName?: string;
  roomCode: string;
  roomNumber: string;
  area?: number;
  roomType?: string;
  roomStatus: 'available' | 'occupied' | 'maintenance' | 'reserved';
  position?: string;
  assetOwnership?: string;
  decoration?: string;
  orientation?: string;
  hasWindow?: boolean;
  hasAC?: boolean;
  rentPrice?: number;
  remark?: string;
  assets?: AssetItem[];
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

// 具体资产项目
export interface AssetItem {
  id: number;
  assetCode: string;
  assetName: string;
  assetType: string;
  purchaseDate?: string;
  assetValue?: number;
  depreciationRate?: number;
  status: 'in_use' | 'idle' | 'maintenance' | 'scrapped';
  roomId?: number;
  roomName?: string;
  userId?: number;
  userName?: string;
  department?: string;
  description?: string;
  images?: string[];
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

// 街道信息
export interface Street {
  id: number;
  name: string;
  code: string;
  districtId: number;
  districtName: string;
}

// 查询参数
export interface AssetQueryParams {
  page?: number;
  pageSize?: number;
  sort?: string;
  order?: 'asc' | 'desc';
  assetName?: string;
  streetId?: number;
  status?: Status;
  assetTags?: string;
}

export interface BuildingQueryParams {
  page?: number;
  pageSize?: number;
  sort?: string;
  order?: 'asc' | 'desc';
  assetId?: number;
  buildingName?: string;
  buildingType?: string;
  status?: Status;
}

export interface FloorQueryParams {
  page?: number;
  pageSize?: number;
  sort?: string;
  order?: 'asc' | 'desc';
  buildingId: number;
  floorName?: string;
  status?: Status;
}

export interface RoomQueryParams {
  page?: number;
  pageSize?: number;
  sort?: string;
  order?: 'asc' | 'desc';
  floorId?: number;
  roomNumber?: string;
  roomType?: string;
  roomStatus?: string;
}

export interface AssetItemQueryParams {
  page?: number;
  pageSize?: number;
  sort?: string;
  order?: 'asc' | 'desc';
  roomId?: number;
  assetName?: string;
  assetType?: string;
  status?: string;
  department?: string;
}

// 表单数据
export interface AssetFormData {
  assetName: string;
  streetId?: number;
  address?: string;
  longitude?: number;
  latitude?: number;
  landNature?: string;
  totalArea?: number;
  rentableArea?: number;
  assetTags?: string[];
  description?: string;
  useDate?: string;
  status: Status;
}

export interface BuildingFormData {
  assetId: number;
  buildingName: string;
  buildingShortName?: string;
  buildingType: string;
  investor?: string;
  location?: string;
  totalArea?: number;
  rentableArea?: number;
  buildingHeight?: number;
  floorCount?: number;
  undergroundFloors?: number;
  constructionYear?: string;
  elevatorCount?: number;
  parkingSpaces?: number;
  greenRate?: number;
  propertyCompany?: string;
  propertyPhone?: string;
  features?: string[];
  description?: string;
  buildDate?: string;
  status: Status;
}

export interface FloorFormData {
  buildingId: number;
  floorName: string;
  floorNumber: number;
  totalArea?: number;
  rentableArea?: number;
  floorType?: string;
  avgRentPrice?: number;
  description?: string;
  status: Status;
}

export interface RoomFormData {
  floorId: number;
  roomNumber: string;
  area?: number;
  roomType?: string;
  roomStatus: 'available' | 'occupied' | 'maintenance' | 'reserved';
  position?: string;
  assetOwnership?: string;
  decoration?: string;
  orientation?: string;
  hasWindow?: boolean;
  hasAC?: boolean;
  rentPrice?: number;
  remark?: string;
}

export interface AssetItemFormData {
  assetName: string;
  assetType: string;
  purchaseDate?: string;
  assetValue?: number;
  depreciationRate?: number;
  status: 'in_use' | 'idle' | 'maintenance' | 'scrapped';
  roomId?: number;
  userId?: number;
  department?: string;
  description?: string;
}

// 资产转移记录
export interface AssetTransferLog {
  id: number;
  assetId: number;
  assetName: string;
  fromRoomId?: number;
  fromRoomName?: string;
  toRoomId?: number;
  toRoomName?: string;
  fromUserId?: number;
  fromUserName?: string;
  toUserId?: number;
  toUserName?: string;
  transferReason: string;
  transferDate: string;
  operatorId: number;
  operatorName: string;
  remark?: string;
  createdAt: string;
}

// 资产盘点记录
export interface AssetInventoryLog {
  id: number;
  inventoryCode: string;
  inventoryName: string;
  inventoryDate: string;
  inventoryType: 'full' | 'partial' | 'spot';
  scopeType: 'all' | 'building' | 'floor' | 'room' | 'department';
  scopeId?: number;
  scopeName?: string;
  totalAssets: number;
  foundAssets: number;
  missingAssets: number;
  excessAssets: number;
  damagedAssets: number;
  status: 'pending' | 'in_progress' | 'completed' | 'cancelled';
  operatorId: number;
  operatorName: string;
  remark?: string;
  createdAt: string;
  updatedAt: string;
}

// 统计数据类型
export interface AssetStatistics {
  totalAssets: number;
  totalArea: number;
  totalValue: number;
  averageArea: number;
  averageValue: number;
  assetsByType: Record<string, number>;
  assetsByStatus: Record<string, number>;
  assetsByProvince: Record<string, number>;
  assetsByCity: Record<string, number>;
  recentAssets: Asset[];
  valueAppreciation: number;
  occupancyRate: number;
  monthlyIncome: number;
  annualIncome: number;
}

export interface BuildingStatistics {
  totalBuildings: number;
  totalArea: number;
  averageArea: number;
  buildingsByType: Record<string, number>;
  buildingsByStatus: Record<string, number>;
  buildingsByStructure: Record<string, number>;
  buildingsByEnergyRating: Record<string, number>;
  averageAge: number;
  occupancyRate: number;
  monthlyIncome: number;
}

export interface FloorStatistics {
  totalFloors: number;
  totalArea: number;
  averageArea: number;
  floorsByType: Record<string, number>;
  floorsByStatus: Record<string, number>;
  averageRoomCount: number;
  occupancyRate: number;
  monthlyIncome: number;
}

export interface RoomStatistics {
  totalRooms: number;
  totalArea: number;
  averageArea: number;
  roomsByType: Record<string, number>;
  roomsByStatus: Record<string, number>;
  roomsByOccupancy: Record<string, number>;
  occupancyRate: number;
  averageRent: number;
  monthlyIncome: number;
  vacancyRate: number;
}

// 地图相关类型
export interface MapPoint {
  id: number;
  name: string;
  type: 'asset' | 'building';
  longitude: number;
  latitude: number;
  address: string;
  status: string;
  area?: number;
  value?: number;
  icon?: string;
  color?: string;
  data?: Asset | Building;
}

export interface MapRegion {
  code: string;
  name: string;
  level: 'province' | 'city' | 'district';
  parent?: string;
  center: [number, number];
  bounds?: [[number, number], [number, number]];
  assetCount: number;
  totalArea: number;
  totalValue: number;
}

// 这些类型已经在前面定义了，移除重复定义

File: backend/internal/model/asset.go
package model

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
)

// Asset 资产模型
type Asset struct {
	AuditModel
	AssetCode    string    `gorm:"uniqueIndex;size:50;not null" json:"asset_code"`  // 资产编码
	AssetName    string    `gorm:"size:100;not null;index" json:"asset_name"`      // 资产名称
	StreetID     uint      `gorm:"index;not null" json:"street_id"`                // 街道ID
	Address      string    `gorm:"size:200" json:"address"`                        // 详细地址
	Longitude    float64   `json:"longitude"`                                       // 经度
	Latitude     float64   `json:"latitude"`                                        // 纬度
	LandNature   string    `gorm:"size:50" json:"land_nature"`                     // 土地性质
	TotalArea    float64   `json:"total_area"`                                      // 总面积(平方米)
	RentableArea float64   `json:"rentable_area"`                                   // 可租赁面积(平方米)
	AssetTags    StringArray `gorm:"type:json" json:"asset_tags"`                    // 资产标签
	Description  string    `gorm:"type:text" json:"description"`                    // 描述
	Status       string    `gorm:"size:20;default:'normal'" json:"status"`          // 状态：normal-正常，disabled-禁用
	Street       *Organization `gorm:"foreignKey:StreetID" json:"street,omitempty"`  // 街道信息
	Buildings    []Building `gorm:"foreignKey:AssetID" json:"buildings,omitempty"`  // 楼宇列表
}

// TableName 设置表名
func (Asset) TableName() string {
	return "t_asset"
}

// Building 楼宇模型
type Building struct {
	AuditModel
	BuildingCode      string     `gorm:"uniqueIndex;size:50;not null" json:"building_code"`      // 楼宇编码
	BuildingName      string     `gorm:"size:100;not null;index" json:"building_name"`          // 楼宇名称
	AssetID           uint       `gorm:"index;not null" json:"asset_id"`                        // 资产ID
	BuildingType      string     `gorm:"size:50" json:"building_type"`                          // 楼宇类型
	TotalFloors       int        `json:"total_floors"`                                           // 总楼层数
	UndergroundFloors int        `json:"underground_floors"`                                     // 地下楼层数
	TotalArea         float64    `json:"total_area"`                                             // 总面积(平方米)
	RentableArea      float64    `json:"rentable_area"`                                          // 可租赁面积(平方米)
	ConstructionYear  string     `gorm:"size:10" json:"construction_year"`                       // 建造年份
	ElevatorCount     int        `json:"elevator_count"`                                         // 电梯数量
	ParkingSpaces     int        `json:"parking_spaces"`                                         // 停车位数量
	GreenRate         float64    `json:"green_rate"`                                             // 绿化率(%)
	PropertyCompany   string     `gorm:"size:100" json:"property_company"`                       // 物业公司
	PropertyPhone     string     `gorm:"size:20" json:"property_phone"`                          // 物业电话
	Features          StringArray `gorm:"type:json" json:"features"`                             // 配套设施
	Description       string     `gorm:"type:text" json:"description"`                           // 描述
	Status            string     `gorm:"size:20;default:'normal'" json:"status"`                 // 状态
	Asset             *Asset     `gorm:"foreignKey:AssetID" json:"asset,omitempty"`             // 资产信息
	Floors            []Floor    `gorm:"foreignKey:BuildingID" json:"floors,omitempty"`          // 楼层列表
}

// TableName 设置表名
func (Building) TableName() string {
	return "t_building"
}

// Floor 楼层模型
type Floor struct {
	AuditModel
	BuildingID       uint      `gorm:"index;not null" json:"building_id"`              // 楼宇ID
	FloorNumber      int       `gorm:"index" json:"floor_number"`                      // 楼层号
	FloorName        string    `gorm:"size:50" json:"floor_name"`                      // 楼层名称
	FloorArea        float64   `json:"floor_area"`                                      // 楼层面积(平方米)
	RentableArea     float64   `json:"rentable_area"`                                   // 可租赁面积(平方米)
	RentedArea       float64   `json:"rented_area"`                                     // 已租赁面积(平方米)
	AvgRentPrice     float64   `json:"avg_rent_price"`                                  // 平均租金(元/平方米/月)
	OccupancyRate    float64   `json:"occupancy_rate"`                                  // 出租率(%)
	Description      string    `gorm:"type:text" json:"description"`                    // 描述
	Status           string    `gorm:"size:20;default:'normal'" json:"status"`          // 状态
	Building         *Building `gorm:"foreignKey:BuildingID" json:"building,omitempty"` // 楼宇信息
	Rooms            []Room    `gorm:"foreignKey:FloorID" json:"rooms,omitempty"`       // 房间列表
}

// TableName 设置表名
func (Floor) TableName() string {
	return "t_floor"
}

// Room 房间模型
type Room struct {
	AuditModel
	FloorID       uint    `gorm:"index;not null" json:"floor_id"`           // 楼层ID
	RoomNumber    string  `gorm:"size:50;index" json:"room_number"`         // 房间号
	RoomType      string  `gorm:"size:50" json:"room_type"`                 // 房间类型：office-办公室，meeting-会议室，other-其他
	RoomArea      float64 `json:"room_area"`                                 // 房间面积(平方米)
	RentPrice     float64 `json:"rent_price"`                                // 租金(元/月)
	Decoration    string  `gorm:"size:50" json:"decoration"`                 // 装修情况：blank-毛坯，simple-简装，luxury-精装
	Orientation   string  `gorm:"size:50" json:"orientation"`                // 朝向：east-东，south-南，west-西，north-北
	HasWindow     bool    `json:"has_window"`                                // 是否有窗
	HasAC         bool    `json:"has_ac"`                                    // 是否有空调
	Description   string  `gorm:"type:text" json:"description"`              // 描述
	Status        string  `gorm:"size:20;default:'available'" json:"status"` // 状态：available-可租，rented-已租，maintenance-维护中
	Floor         *Floor  `gorm:"foreignKey:FloorID" json:"floor,omitempty"` // 楼层信息
}

// TableName 设置表名
func (Room) TableName() string {
	return "t_room"
}

// Street 街道模型（组织表的视图）
type Street struct {
	ID   uint   `json:"id"`
	Name string `json:"name"`
	Code string `json:"code"`
}

// StringArray 自定义类型，用于存储JSON数组
type StringArray []string

// Value 实现driver.Valuer接口
func (s StringArray) Value() (driver.Value, error) {
	if s == nil {
		return "[]", nil
	}
	return json.Marshal(s)
}

// Scan 实现sql.Scanner接口
func (s *StringArray) Scan(value interface{}) error {
	if value == nil {
		*s = []string{}
		return nil
	}
	
	switch v := value.(type) {
	case []byte:
		return json.Unmarshal(v, s)
	case string:
		return json.Unmarshal([]byte(v), s)
	default:
		return fmt.Errorf("cannot scan type %T into StringArray", value)
	}
}